### ISSUE TITLE
Tensorboard monitor for recent tensorflow and pytorch

### ISSUE THREAD

**Issue by jungwon-choi** [guochengqian/TENet#12]
I attached the code for the guys who want to use recent tensorflow >= 2.x

    import tensorflow as tf
    import numpy as np

    try:
        from StringIO import StringIO  # Python 2.7
    except ImportError:
        from io import BytesIO         # Python 3.x

    class Tf_Logger(object):
    
        def __init__(self, log_dir):
            """Create a summary writer logging to log_dir."""
            self.writer = tf.summary.create_file_writer(log_dir)

        def scalar_summary(self, tag, value, step):
            """Log a scalar variable."""
            with self.writer.as_default():
                tf.summary.scalar(tag, value, step)
                self.writer.flush()

        def image_summary(self, tag, images, step):
            """Log a list of images."""
            with self.writer.as_default():
                for i, img in enumerate(images):
                    if len(img.shape) == 2:
                        img = img[np.newaxis,:,:,np.newaxis]
                    elif len(img.shape) == 3:
                        img = img[np.newaxis,:,:,:]
                    tf.summary.image('%s/%d' % (tag, i), img, step)
                self.writer.flush()
        
        def histo_summary(self, tag, values, step, bins=1000):
            """Log a histogram of the tensor of values."""

            # Create a histogram using numpy
            counts, bin_edges = np.histogram(values, bins=bins)

            # Fill the fields of the histogram proto
            hist = tf.HistogramProto()
            hist.min = float(np.min(values))
            hist.max = float(np.max(values))
            hist.num = int(np.prod(values.shape))
            hist.sum = float(np.sum(values))
            hist.sum_squares = float(np.sum(values**2))

            # Drop the start of the first bin
            bin_edges = bin_edges[1:]

            # Add bin edges and counts
            for edge in bin_edges:
                hist.bucket_limit.append(edge)
            for c in counts:
                hist.bucket.append(c)

            # Create and write Summary
            with self.writer.as_default():
                tf.summary.histogram(tag, hist, step)
                self.writer.flush()

**Comment by CharlesJWC** [comment]
plus you have to edit each train.py as below 
from

    info = {
                    'loss': loss,
                    'sr_loss': sr_loss,
                    'dm_loss': dm_loss,
                    'psnr': cur_psnr
               }

to
info = {
                'loss': loss.item(),
                'sr_loss': sr_loss.item(),
                'dm_loss': dm_loss.item(),
                'psnr': cur_psnr
            }

**Comment by CharlesJWC** [comment]
However, I found that tensorflow requires a lot of memory just to log the losses.
So, I changed the code with tensorboard for pytorch.

**Comment by CharlesJWC** [comment]
In this case, you don't need to transpose RGB images at valid() function.
   # show images
    img = img[0:1, 0:4]
    img = torch.clamp(img.cpu()[0, 0], 0, 1).detach().numpy()
    gt = torch.clamp(gt.cpu()[0], 0, 1).detach().numpy()
    # gt = np.transpose(gt, [1, 2, 0]) 	# permute
    output = torch.clamp(output.cpu()[0], 0, 1).detach().numpy()
    # output = np.transpose(output, [1, 2, 0])

**Comment by guochengqian** [comment]
@jungwon-choi  thank you for your updates! Really helpful.
